cl/redisapp/leaderfollower/leaderfollower.go

```diff
@@ -0,0 +1,323 @@
+						lfm.logger.Error("Leader: failed to reach geth node after max attempts, exiting")
+						err := lfm.leaderProc.Stop()
+						if err != nil {
+							lfm.logger.Error("Leader: Failed to stop leader election", "error", err)
+						}
+						lfm.blockBuilder.LastCallTime = time.Time{}
+					}
+				}
+			} else {
+				if !lfm.isFollowerInitialized.Load() {
+					if err := lfm.blockBuilder.ProcessLastPayload(ctx); err != nil {
+						lfm.logger.Error("Error processing last payload", "error", err)
+						continue
+					}
+					lfm.isFollowerInitialized.Store(true)
+				}
+				if err := lfm.followerWork(ctx); err != nil {
+					lfm.logger.Error("Error in follower work", "error", err)
+				}
+
+				// will do nothing, in case election already started
+				// needed, in case leader stopped being a leader bcs of geth issue
+				lfm.leaderProc.Start()
+			}
+		}
+	}
+}
+
+func (lfm *LeaderFollowerManager) leaderWork(ctx context.Context) error {
+	lfm.logger.Info("Leader: Performing leader tasks")
+
+	// Ensure state is synchronized before starting leader tasks
+	isSync, err := lfm.ensureStateSynchronized(ctx)
+	if err != nil {
+		lfm.logger.Error("Leader: Failed to synchronize state", "error", err)
+		return err
+	}
+
+	// if not sync, return for the signal to be a leader
+	if !isSync {
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			lfm.logger.Debug("Leader: Exiting")
+			return nil
+		default:
+			bbState := lfm.stateManager.GetBlockBuildState(ctx)
+			currentStep := bbState.CurrentStep
+
+			switch currentStep {
+			case types.StepBuildBlock:
+				lfm.logger.Info("Leader: StepBuildBlock")
+				if err := lfm.blockBuilder.GetPayload(ctx); err != nil {
+					lfm.logger.Error("Leader: GetPayload failed", "error", err)
+					if resetErr := lfm.stateManager.ResetBlockState(ctx); resetErr != nil {
+						lfm.logger.Error("Leader: Failed to reset block state", "error", resetErr)
+					}
+
+					return err
+				}
+			case types.StepFinalizeBlock:
+				lfm.logger.Info("Leader: StepFinalizeBlock")
+				if err := lfm.blockBuilder.FinalizeBlock(ctx, bbState.PayloadID, bbState.ExecutionPayload, ""); err != nil {
+					lfm.logger.Error("Leader: FinalizeBlock failed", "error", err)
+					if resetErr := lfm.stateManager.ResetBlockState(ctx); resetErr != nil {
+						lfm.logger.Error("Leader: Failed to reset block state", "error", resetErr)
+					}
+					return err
+				}
+				if err := lfm.stateManager.ResetBlockState(ctx); err != nil {
+					lfm.logger.Error("Leader: Failed to reset block state", "error", err)
+					return err
+				}
+			default:
+				lfm.logger.Warn("Leader: Unknown current step", "current_step", currentStep.String())
+				if err := lfm.stateManager.ResetBlockState(ctx); err != nil {
+					lfm.logger.Error("Leader: Failed to reset block state", "error", err)
+					return err
+				}
+			}
+		}
+	}
+}
+
+func (lfm *LeaderFollowerManager) ensureStateSynchronized(ctx context.Context) (bool, error) {
```

Comment:
In your case, there could be a situation where the leader node is out of sync. The entire network will stop generating blocks because all nodes will be waiting for the leader to finally sync.